### PR TITLE
[#2416] feat(trino-connector): Support system table catalog

### PR DESCRIPTION
**Pull request by diqiu50**
### What changes were proposed in this pull request?

Support the system table catalog to display all loaded catalog information in the Trino connector.

### Why are the changes needed?

Feature: #2416 

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

Add new UT

### CODE REVIEW

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConnectorFactory.java [hunk 1]
```diff
@@ -55,7 +58,7 @@ public Connector create(
           CatalogConnectorFactory catalogConnectorFactory =
               new CatalogConnectorFactory(catalogInjector);
 
-          catalogConnectorManager =
+          this.catalogConnectorManager =
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConnectorFactory.java:61]: It's not advised to use `this` out of the constructor.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/GravitinoSystemConnector.java [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+import java.lang.invoke.MethodHandles;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * GravitinoSystemConnector is primarily used to drive the GravitinoCatalogManager to load catalog
+ * connectors managed in the Gravitino server. After users configure the Gravitino connector through
+ * Trino catalog configuration, a DummyGravitinoFConnector is initially created. And it provides
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/GravitinoSystemConnector.java:49]: typo:
DummyGravitinoFConnector

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/GravitinoSystemConnector.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    @Override
+    public boolean isRemotelyAccessible() {
+      return true;
+    }
+
+    @Override
+    public List<HostAddress> getAddresses() {
+      return List.of(HostAddress.fromParts("127.0.0.1", 8080));
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/GravitinoSystemConnector.java:195]: Why do you fixed the port to `8080`? What if users change the default port?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java [hunk 4]
```diff
@@ -174,15 +174,19 @@ public void loadCatalogs(GravitinoMetaLake metalake) {
...
-                  GravitinoCatalog gravitinoCatalog = new GravitinoCatalog(catalog);
-                  CatalogConnectorContext catalogConnectorContext =
-                      catalogConnectorFactory.loadCatalogConnector(
-                          nameIdentifier, metalake, gravitinoCatalog);
-
-                  catalogConnectors.put(catalogName, catalogConnectorContext);
-                  catalogInjector.injectCatalogConnector(catalogName);
-                  LOG.info(
-                      "Load catalog {} in metalake {} successfully.", catalogName, metalake.name());
+                  if (catalog.type() == Catalog.Type.RELATIONAL) {
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:177]: So we would not support fileset and stream catalog any more?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  @Override
+  public Page loadPageData() {
+    int size = catalogConnectorManager.getCatalogs().size();
+
+    BlockBuilder nameColumnBuilder = VARCHAR.createBlockBuilder(null, size);
+    BlockBuilder providerColumnBuilder = VARCHAR.createBlockBuilder(null, size);
+    MapBlockBuilder propertyColumnBuilder = STRING_MAPTYPE.createBlockBuilder(null, size);
+
+    for (String catalogName : catalogConnectorManager.getCatalogs()) {
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java:54]: Why not directly add a method `getCatalogConnectors` in `CatalogConnectorManager`.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java (hunk 4 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorManager.java:177]: Yes

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/GravitinoSystemConnector.java (hunk 3 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/GravitinoSystemConnector.java:195]: The address is never used by Trino. I will remove it

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConnectorFactory.java [hunk 6]
```diff
@@ -81,15 +86,16 @@ public Connector create(
       Preconditions.checkNotNull(catalogConnectorContext, "catalogConnector is not null");
       return catalogConnectorContext.getConnector();
     } else {
-      // The static connector is an instance of DummyGravitinoConnector. It is loaded by Trino using
+      // The static connector is an instance of GravitinoSystemConnector. It is loaded by Trino
+      // using
       // the connector configuration.
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/GravitinoConnectorFactory.java:91]: Format the comment

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorContext.java [hunk 7]
```diff
@@ -97,22 +103,21 @@ Builder withMetalake(GravitinoMetaLake metalake) {
...
       return this;
     }
 
     CatalogConnectorContext build() {
-      Preconditions.checkArgument(catalogName != null, "catalogName is not null");
       Preconditions.checkArgument(metalake != null, "metalake is not null");
       Preconditions.checkArgument(internalConnector != null, "internalConnector is not null");
-      return new CatalogConnectorContext(
-          catalogName, metalake, internalConnector, connectorAdapter);
+      Preconditions.checkArgument(catalog != null, "catalog  ss not null");
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/CatalogConnectorContext.java:119]: What's the meaning of `ss`?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+  private final CatalogConnectorManager catalogConnectorManager;
+
+  public GravitinoSystemTableCatalog(CatalogConnectorManager catalogConnectorManager) {
+    this.catalogConnectorManager = catalogConnectorManager;
+  }
+
+  @Override
+  public Page loadPageData() {
+    int size = catalogConnectorManager.getCatalogs().size();
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java:48]: I think you can make `catalogConnectorManager.getCatalogs()` as a temporary variant, then you can reuse it in line 54.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+import io.trino.spi.connector.SchemaTableName;
+import io.trino.spi.type.MapType;
+import io.trino.spi.type.TypeOperators;
+import java.util.List;
+
+/** An implementation of the catalog system table * */
+public class GravitinoSystemTableCatalog extends GravitinoSystemTable {
+
+  public static final SchemaTableName TABLE_NAME =
+      new SchemaTableName(SYSTEM_TABLE_SCHEMA_NAME, "catalog");
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java:28]: Is the name of the system catalog `catalog`? I would suggest using `default`, `catalog` seems to be very strange.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableFactory.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import io.trino.spi.connector.ConnectorTableMetadata;
+import io.trino.spi.connector.SchemaTableName;
+import java.util.HashMap;
+import java.util.Map;
+
+/* This class managed all the system tables */
+public class GravitinoSystemTableFactory {
+
+  private final CatalogConnectorManager catalogConnectorManager;
+  public static Map<SchemaTableName, GravitinoSystemTable> systemTableDefines = new HashMap<>();
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableFactory.java:19]: If the reference of `systemTableDefines` will not be changed, I would suggest adding a `final` flag.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java (hunk 9 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java:28]: Actually, the table name is `catalog`. The meaning  of `default` is not clear.
Users usually query with the table like this:
```
select * from graviton.system.catalog
```

Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java:28]: So the schema name is `system` and table name is `catalog`?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  public Page loadPageData() {
+    List<GravitinoCatalog> catalogs = catalogConnectorManager.getCatalogs();
+    int size = catalogs.size();
+
+    BlockBuilder nameColumnBuilder = VARCHAR.createBlockBuilder(null, size);
+    BlockBuilder providerColumnBuilder = VARCHAR.createBlockBuilder(null, size);
+    MapBlockBuilder propertyColumnBuilder = STRING_MAPTYPE.createBlockBuilder(null, size);
+
+    for (GravitinoCatalog catalog : catalogs) {
+      Preconditions.checkNotNull(catalog, "catalog not be null");
```
Comment (yuqi1129) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java:56]: catalog can't be null/ catalog should not be null.

**Comment by yuqi1129** [comment]
Overall, LGTM except for a few minor issues, please resolve them and let me know.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java (hunk 11 above)
Comment (diqiu50) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/system/table/GravitinoSystemTableCatalog.java:56]: Yes, schema name is system and table name is catalog

**Comment by yuqi1129** [comment]
LGTM

**Review by yuqi1129 — APPROVED** [review]

**Comment by yuqi1129** [comment]
@jerryshao 
Do you need to take a look?

**Comment by jerryshao** [comment]
You can go ahead if you guys feel confident. @yuqi1129 @diqiu50

**Pull request merged**